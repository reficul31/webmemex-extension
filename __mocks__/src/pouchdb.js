import PouchDB from 'pouchdb-node'
import PouchDBMemory from 'pouchdb-adapter-memory'

PouchDB.plugin(PouchDBMemory)

const pouchdbOptions = {
    name: 'testdb',
    auto_compaction: true,
    adapter: 'memory',
}

const db = PouchDB(pouchdbOptions)
export default db
