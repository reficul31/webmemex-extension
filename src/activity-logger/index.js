// Stuff that is to be accessible from other modules (folders)

import docuri from 'docuri'
import randomString from 'src/util/random-string'

export const visitKeyPrefix = 'visit/'

/**
* Creates an _id string given the variables, or vice versa parses such strings
* We simply use the creation time for the id, for easy chronological sorting.
* We add a random string we call a 'nonce' to prevent accidental collisions.
*
* @param {Number} timestamp - timestamp of the document
* @param {Number} nonce - random 10 digit number
*
* @returns {String} - string id of the document
*/
export const convertVisitDocId = docuri.route(`${visitKeyPrefix}:timestamp/:nonce`)

/**
* Extract the type, timestamp and nonce from an ID of a document
*
* @param {String} - string id of the document
*/
const convertAnyTimestampedDocId = docuri.route(':type/:timestamp/:nonce')
export const getTimestamp = doc =>
    Number.parseInt(convertAnyTimestampedDocId(doc._id).timestamp)

/**
* Generate a visit document ID from the timestamp and a random string
*
* @param {Number} timestamp - timestamp of the visit document
* @param {Number} nonce - random 10 digit number
*
* @returns {string} - document id of the visit document
*/
export function generateVisitDocId({timestamp, nonce} = {}) {
    const date = timestamp ? new Date(timestamp) : new Date()
    return convertVisitDocId({
        timestamp: date.getTime(),
        nonce: nonce || randomString(),
    })
}

/**
* Tell whether a page can be stored
*
* @param {string} url - url of the visit document
*
* @returns {boolean} - whether or not the page is loggable
*/
export function isLoggable({url}) {
    // Only http(s) pages. Ignoring data URLs, newtab, ...
    const loggableUrlPattern = /^https?:\/\//
    return loggableUrlPattern.test(url)
}

/**
* Tell whether a page should be stored
*
* @param {string} url - url of the visit document
*
* @returns {boolean} - whether or not the page should be stored
*/
export function shouldBeLogged({url}) {
    // Currently, we log everything that we think we can log.
    // TODO Add ignore-/blacklist lookup here (issue #22, #94)
    return isLoggable({url})
}
