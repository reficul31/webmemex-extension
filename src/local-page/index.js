import get from 'lodash/fp/get'

/**
* Extract the local page URL from the page instance.
*
* @param {Page} page - stored page in the database
*
* @returns {string} - URL used to refer to the local page
*/
export function hrefForLocalPage({page}) {
    // Check if it has a stored page attached at all.
    if (!get(['_attachments', 'frozen-page.html'])(page)) {
        return undefined
    }
    const pageId = page._id
    const url = `/local-page.html?page=${encodeURIComponent(pageId)}`
    const hash = (page.url && page.url.split('#')[1])
    const href = (hash !== undefined) ? url + '#' + hash : url
    return href
}
