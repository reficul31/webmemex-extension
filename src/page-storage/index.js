import docuri from 'docuri'
import randomString from 'src/util/random-string'

export const pageKeyPrefix = 'page/'

export const convertPageDocId = docuri.route(`${pageKeyPrefix}:timestamp/:nonce`)

/**
* Generate a page doc id
*
* @param {Number} timestamp - timestamp of the page
* @param {string} nonce - random string of 10 digit number
*
* @returns {string} - doc id of the page as a string
*/
export function generatePageDocId({timestamp, nonce} = {}) {
    const date = timestamp ? new Date(timestamp) : new Date()
    return convertPageDocId({
        timestamp: date.getTime(),
        nonce: nonce || randomString(),
    })
}
