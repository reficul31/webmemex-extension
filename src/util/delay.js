/**
* Creates a delay of the timeperiod specified
*
* @param {Number} t - timeperiod for the delay
*
* @returns {Promise} - resolves after a timeout
*/
export default function delay(t) {
    return new Promise(
        resolve => window.setTimeout(resolve, t)
    )
}
