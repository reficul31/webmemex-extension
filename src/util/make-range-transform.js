import clamp from 'lodash/fp/clamp'
import compose from 'lodash/fp/compose'

/**
* Create an affine transformation to map numbers from one range to another.
*
* @param {Array} domain - Domain range of values for input
* @param {Array} range - Range of values for mapping
* @param {boolean} clampOutput - whether the values above or below the range be capped or not
*
* @returns {Function} - range transform function
*/
export function makeRangeTransform({domain: [minX, maxX], range: [minY, maxY], clampOutput = false}) {
    const scale = (maxY - minY) / (maxX - minX)
    const bias = minY - minX * scale
    const transform = x => bias + x * scale
    return clampOutput ? compose(clamp(minY, maxY), transform) : transform
}

/**
* Create a transformation that passes the input through a monotonic nonlinearity
*
* @param {Array} domain - Domain range of values for input
* @param {Array} range - Range of values for mapping
* @param {Function} nonlinearity - non linearity function
* @param {boolean} clampOutput - whether the values above or below the range be capped or not
*
* returns {function} - non linear transform function
*/
export function makeNonlinearTransform({domain, range, nonlinearity, clampOutput = false}) {
    return function nonLinearTransform(input) {
        const toOutputRange = makeRangeTransform({
            domain: domain.map(nonlinearity),
            range,
            clampOutput,
        })
        return toOutputRange(nonlinearity(input))
    }
}
