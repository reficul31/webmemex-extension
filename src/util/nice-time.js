/**
* Returns the hour as a string
*
* @param {Date} date - date to be processed
*
* @returns {string} - hour of the Date as a string
*/
function hourString(date) {
    // return date.toLocaleTimeString([], {hour: 'numeric', minute: '2-digit'})
    return `${('0' + date.getHours()).slice(-2)}:${('0' + date.getMinutes()).slice(-2)}`
}

/**
* Returns the day as a string
*
* @param {Date} date - date to be processed
*
* @returns {string} - day of the Date as a string
*/
function dayString(date) {
    const days = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat']
    return days[date.getDay()]
}

/**
* Returns the month as a string
*
* @param {Date} date - date to be processed
*
* @returns {string} - month of the Date as a string
*/
function monthString(date) {
    const months = ['Jan', 'Feb', 'March', 'Apr', 'May', 'June', 'July', 'Aug', 'Sept', 'Oct', 'Nov', 'Dec']
    return months[date.getMonth()]
}

/**
* Used to format the time format to get a readable string
*
* @param {Date} date - date to be processed
* @param {Date} now - date from where the difference is to be measured
*
* @returns {string} - a human redable formatted string for the time of the Date instance from now
*/
export default function niceTime(date, {now = undefined} = {}) {
    const then = new Date(date)
    now = now || new Date()
    const secondsAgo = (now - then) / 1000
    if (secondsAgo < -60 * 60 * 24) {
        return 'in the future?!'
    }
    if (secondsAgo < -60) {
        return 'soon?!'
    }
    if (secondsAgo < 90) { return 'now' }
    if (secondsAgo < 60 * 10) { return `${Math.round(secondsAgo / 60)} minutes ago` }
    if (secondsAgo < 60 * 60) { return `${Math.round(secondsAgo / 60 / 5) * 5} minutes ago` }
    if (secondsAgo < 60 * 60 * 24
        && (now.getDay() === then.getDay() || secondsAgo < 60 * 60 * 6)) { return hourString(then) }
    if (secondsAgo < 60 * 60 * 24) { return `Yesterday ${hourString(then)}` }
    if (secondsAgo < 60 * 60 * 24 * 3) { return `${dayString(then)} ${hourString(then)}` }
    if (then.getYear() === now.getYear()) { return `${then.getDate()} ${monthString(then)}` }
    return `${then.getDate()} ${monthString(then)} ${then.getFullYear()}`
}

/**
* Used to format the date format to get a readable string
*
* @param {Date} date - date to be processed
* @param {Date} now - date from where the difference is to be measured
*
* @returns {string} - a human redable formatted string for the date of the Date instance from now
*/
export function niceDate(date, {now = new Date()} = {}) {
    const then = new Date(date)
    if (then.getYear() !== now.getYear()) {
        return `${then.getDate()} ${monthString(then)} ${then.getFullYear()}`
    }
    if (then.getMonth() === now.getMonth()) {
        const daysAgo = now.getDate() - then.getDate()
        if (daysAgo === 0) return `Today`
        if (daysAgo === 1) return `Yesterday`
    }
    return `${dayString(then)} ${then.getDate()} ${monthString(then)}`
}
