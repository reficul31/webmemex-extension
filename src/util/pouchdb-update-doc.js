import assocPath from 'lodash/fp/assocPath'


// Set a limit to not hang forever if something is more fundamentally broken.
const MAX_ATTEMPTS = 9999

/**
* Function to update the page stub created
*
* @param {} db - database instance of pouchdb
* @param {string} docId - document id to be updated
* @param {function} updateFunc - update function used to perform the changes on the given doc
*/
async function updateDoc(db, docId, updateFunc) {
    let conflict
    let attemptCount = 0
    do {
        const doc = await db.get(docId)
        const updatedDoc = await updateFunc(doc)
        try {
            await db.put(updatedDoc)
            conflict = false
        } catch (err) {
            if (err.name === 'conflict' && ++attemptCount < MAX_ATTEMPTS) {
                // Try again.
                conflict = true
            } else {
                throw err
            }
        }
    } while (conflict)
}
export default updateDoc

/**
* A shorthand for adding/updating an attachment to an existing document.
*
* @param {} db - database instance of pouchdb
* @param {string} docId - document id to be updated
* @param {string} attachmentId - string used to denote the attachment type
* @param {Blob} blob - object to be attached to the document
*/
export async function setAttachment(db, docId, attachmentId, blob) {
    await updateDoc(db, docId,
        doc => assocPath(
            ['_attachments', attachmentId],
            {content_type: blob.type, data: blob}
        )(doc)
    )
}
