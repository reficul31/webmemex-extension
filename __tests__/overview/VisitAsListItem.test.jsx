import React from 'react'
import VisitAsListItem from 'src/overview/components/VisitAsListItem'
import renderer from 'react-test-renderer'
import * as row from './row.data.json'
import * as niceTimeModule from 'src/util/nice-time'
import * as pageViewer from 'src/page-viewer'

describe('VisitAsListItem snapshot tests', () => {
    test('should test against the given snapshot', () => {
        const spy = jest.spyOn(niceTimeModule, 'default')
                    .mockReturnValueOnce('20 minutes ago')
        const component = renderer.create(
            <VisitAsListItem doc={row.doc} compact={false} />
            )
        let tree = component.toJSON()
        expect(tree).toMatchSnapshot()
        spy.mockRestore()
    })

    test('should call the niceTime util', () => {
        const spy = jest.spyOn(niceTimeModule, 'default')
        renderer.create(
            <VisitAsListItem doc={row.doc} compact={false} />
            )
        expect(spy).toHaveBeenCalledTimes(1)
        spy.mockRestore()
    })

    test('should check for the localVersion', () => {
        const spy = jest.spyOn(pageViewer, 'localVersionAvailable')
        renderer.create(
            <VisitAsListItem doc={row.doc} compact={false} />
            )
        expect(spy).toHaveBeenCalledTimes(1)
        spy.mockRestore()
    })
})
