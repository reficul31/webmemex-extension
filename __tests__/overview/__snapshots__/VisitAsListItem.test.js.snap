// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VisitAsListItem snapshot tests should test against the given snapshot 1`] = `
<a
  className="root"
  href="https://www.google.co.in/?gfe_rd=cr&ei=zrsdWc2dH9KnvwTejJOIBQ"
  onClick={[Function]}
>
  <div
    className="screenshotContainer"
  >
    <img
      className="favIcon"
      src="img/null-icon.png"
    />
  </div>
  <div
    className="descriptionContainer"
  >
    <div
      className="title"
      title="Google"
    >
      Google
    </div>
    <div
      className="url"
    >
      https://www.google.co.in/?gfe_rd=cr&ei=zrsdWc2dH9KnvwTejJOIBQ
    </div>
    <div
      className="time"
    >
      20 minutes ago
    </div>
  </div>
  <div
    className="buttonsContainer"
  >
    <img
      alt="ðŸ—‘ remove"
      src="img/trash-icon.png"
    />
  </div>
</a>
`;
